File: ts/tools/parity-roundtrip.js
// parity-roundtrip — read a crtx v0.1 envelope JSON file at argv[2],
// parse via @hop-top/stem parseEnvelope, re-serialize via
// serializeEnvelope, write to stdout. Used by tools/parity/runner.sh.
//
// Implementation is plain JS so it runs without a TS compiler step.
// It depends on the compiled output in ts/dist/ — the runner builds
// the SDK first via `pnpm build` (or `pnpm install --ignore-scripts`
// + `pnpm build`).

const fs = require("node:fs");
const path = require("node:path");

const stem = require(path.resolve(__dirname, "..", "dist", "index.js"));

const argv = process.argv.slice(2);
if (argv.length !== 1) {
  process.stderr.write("usage: parity-roundtrip <envelope.json>\n");
  process.exit(2);
}

const fixturePath = argv[0];
let text;
try {
  text = fs.readFileSync(fixturePath, "utf8");
} catch (err) {
  process.stderr.write(`parity-roundtrip: read ${fixturePath}: ${err.message}\n`);
  process.exit(1);
}

let env;
try {
  env = stem.parseEnvelope(text);
} catch (err) {
  process.stderr.write(`parity-roundtrip: parse: ${err.message}\n`);
  process.exit(1);
}

let out;
try {
  out = stem.serializeEnvelope(env);
} catch (err) {
  process.stderr.write(`parity-roundtrip: serialize: ${err.message}\n`);
  process.exit(1);
}

process.stdout.write(out);

File: rs/examples/parity-roundtrip.rs
//! parity-roundtrip — read a crtx v0.1 envelope JSON file at argv[1],
//! parse via `hop_top_stem::parse_envelope`, re-serialize via
//! `hop_top_stem::serialize_envelope`, write to stdout.
//!
//! Used by `tools/parity/runner.sh`. Run from `rs/` as:
//!
//! ```text
//! cargo run --example parity-roundtrip -- <path>
//! ```

use std::env;
use std::fs;
use std::io::{self, Write};
use std::process::ExitCode;

use hop_top_stem::{parse_envelope, serialize_envelope};

fn main() -> ExitCode {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        eprintln!("usage: parity-roundtrip <envelope.json>");
        return ExitCode::from(2);
    }

    let path = &args[1];
    let text = match fs::read_to_string(path) {
        Ok(t) => t,
        Err(err) => {
            eprintln!("parity-roundtrip: read {}: {}", path, err);
            return ExitCode::from(1);
        }
    };

    let env = match parse_envelope(&text) {
        Ok(e) => e,
        Err(err) => {
            eprintln!("parity-roundtrip: parse: {}", err);
            return ExitCode::from(1);
        }
    };

    let out = match serialize_envelope(&env) {
        Ok(s) => s,
        Err(err) => {
            eprintln!("parity-roundtrip: serialize: {}", err);
            return ExitCode::from(1);
        }
    };

    let stdout = io::stdout();
    let mut handle = stdout.lock();
    if let Err(err) = handle.write_all(out.as_bytes()) {
        eprintln!("parity-roundtrip: write: {}", err);
        return ExitCode::from(1);
    }
    ExitCode::SUCCESS
}
